### PR TITLE
Add Support For SkAdN + Refactor Split Imps

### PR DESCRIPTION
**Pull request by SyntaxNode**
- Implements https://github.com/prebid/prebid-server/issues/1654.
- Refactors splitImps, which became overly complicated when support for imp.ext.prebid.bidder.BIDDER was implemented.
- Added missing unit test coverage for an important aspect of the core logic.

### CODE REVIEW

File: exchange/exchangetest/firstpartydata-imp-ext-multiple-prebid-bidders.json  [hunk 1]
```diff
@@ -61,7 +61,6 @@
                             "bidder": {
                                 "placementId": 1
                             },
-                            "prebid": {},
```
Comment (SyntaxNode) [inline exchange/exchangetest/firstpartydata-imp-ext-multiple-prebid-bidders.json :64]: Previously, `splitImps` would keep an empty prebid json section. The refactor removes it entirely now. I thought that would be a fine change. Do you agree? .. or should I keep the older functionality?

File: exchange/utils.go [hunk 2]
```diff
@@ -327,106 +328,99 @@ func extractBuyerUIDs(user *openrtb.User) (map[string]string, error) {
...
-
-	var prebidExt map[string]json.RawMessage
-	var errs []error
-
-	if err := json.Unmarshal(rawPrebidExt, &prebidExt); err == nil {
-		// Remove the entire bidder field. We will already have the content we need in bidderExts. We
-		// don't want to include other demand partners' bidder params in the sanitized imp.
-		if _, hasBidderField := prebidExt["bidder"]; hasBidderField {
-			delete(prebidExt, "bidder")
+func createSanitizedImpExt(impExt, impExtPrebid map[string]json.RawMessage) (map[string]json.RawMessage, error) {
```
Comment (bsardo) [inline exchange/utils.go:375]: This refactor is definitely an improvement, however it is still a little confusing that this function takes the unmarshaled and marshaled versions of `imp.ext.prebid` ignoring the marshaled version. Perhaps it's not worth considering alternative approaches at this point. I imagine this complexity is going to go away anyway when we complete the request memoization wrapper.

**Review by bsardo — COMMENTED** [review]
This is looking good; just a couple of comments.

File: endpoints/openrtb2/auction.go [hunk 3]
```diff
@@ -830,19 +830,15 @@ func (deps *endpointDeps) validateImpExt(imp *openrtb.Imp, aliases map[string]st
...
-	// Also accept bidder exts within imp[...].ext.prebid.bidder
-	// NOTE: This is not part of the official API yet, so we are not expecting clients
-	// to migrate from imp[...].ext.${BIDDER} to imp[...].ext.prebid.bidder.${BIDDER}
-	// at this time
-	// https://github.com/prebid/prebid-server/pull/846#issuecomment-476352224
-	if rawPrebidExt, ok := bidderExts[openrtb_ext.PrebidExtKey]; ok {
+	// Prefer bidder params from request.imp.ext.prebid.bidder.BIDDER over request.imp.ext.BIDDER
+	// to avoid confusion beteween prebid specific adapter config and other ext protocols.
+	if extPrebidJSON, ok := bidderExts[openrtb_ext.PrebidExtKey]; ok {
 		var prebidExt openrtb_ext.ExtImpPrebid
```
[the remark this answers is not in the document]
Comment (SyntaxNode) [reply endpoints/openrtb2/auction.go:836]: Sure.

File: exchange/utils_test.go [hunk 4]
```diff
@@ -55,6 +55,370 @@ func assertReq(t *testing.T, bidderRequests []BidderRequest,
...
+			givenImpExt:              map[string]json.RawMessage{},
+			givenImpExtPrebidBidders: map[string]json.RawMessage{"bidderA": bidderAJSON, "bidderB": bidderBJSON},
+			expected:                 map[string]json.RawMessage{"bidderA": bidderAJSON, "bidderB": bidderBJSON},
+		},
+		{
+			description:              "Special Names Not Treated Differently -imp.ext.prebid.bidder.BIDDER",
+			givenImpExt:              map[string]json.RawMessage{},
+			givenImpExtPrebidBidders: map[string]json.RawMessage{"prebid": json.RawMessage(`{"prebid":"value1"}}`), "context": json.RawMessage(`{"firstPartyData":"value2"}}`), "skadn": json.RawMessage(`{"skAdNetwork":"value3"}}`)},
+			expected:                 map[string]json.RawMessage{"prebid": json.RawMessage(`{"prebid":"value1"}}`), "context": json.RawMessage(`{"firstPartyData":"value2"}}`), "skadn": json.RawMessage(`{"skAdNetwork":"value3"}}`)},
+		},
```
[the remark this answers is not in the document]
Comment (SyntaxNode) [reply exchange/utils_test.go:401]: Yes, this is the correct behavior. The problem with using imp.ext.BIDDERS is the conflicts with standardized OpenRTB extensions where fields such as imp.ext.skadn have a standard meaning. The names are only reserved due to this conflict. The imp.ext.prebid.bidder.BIDDER has no such conflicts, thus no reserved names. If there is a imp.ext.prebid.bidder.skadn specified then PBS would assume this is a bidder named skadn.

File: exchange/utils_test.go [hunk 5]
```diff
@@ -55,6 +55,370 @@ func assertReq(t *testing.T, bidderRequests []BidderRequest,
...
+	testCases := []struct {
+		description   string
+		givenImps     []openrtb.Imp
+		expectedImps  map[string][]openrtb.Imp
+		expectedError string
+	}{
+		{
+			description:   "Nil",
+			givenImps:     []openrtb.Imp{},
+			expectedImps:  map[string][]openrtb.Imp{},
```
Comment (mansinahar) [inline exchange/utils_test.go:68]: This test case seems to be the same as the next one. Did you mean to assign `nil` value to `givenImps` here?

**Review by mansinahar — COMMENTED** [review]
LGTM except for the minor comment about the test case

File: exchange/utils_test.go (hunk 5 above)
Comment (SyntaxNode) [reply exchange/utils_test.go:68]: Yes, I did. Thank you. Fixed.

**Review by mansinahar — APPROVED** [review]

**Review by bsardo — APPROVED** [review]
LGTM

**Pull request merged**